rust/libnewsboat/src/fslock.rs

```diff
@@ -0,0 +1,97 @@
+use crate::logger::{self, Level};
+use std::fs::{self, File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::io::AsRawFd;
+use std::path::{Path, PathBuf};
+use std::process;
+
+fn remove_lock(lock_path: &Path) {
+    fs::remove_file(lock_path).ok();
+    log!(
+        Level::Debug,
+        "FsLock: removed lockfile {}",
+        lock_path.display()
+    );
+}
+
+#[derive(Default)]
+pub struct FsLock {
+    lock_path: PathBuf,
+    lock_file: Option<File>,
+}
+
+impl Drop for FsLock {
+    fn drop(&mut self) {
+        if self.lock_file.is_some() {
+            remove_lock(&self.lock_path);
+        }
+    }
+}
+
+impl FsLock {
+    pub fn try_lock(&mut self, new_lock_path: &Path, pid: &mut libc::pid_t) -> bool {
+        if self.lock_file.is_some() && self.lock_path == new_lock_path {
+            return true;
+        }
+
+        // pid == 0 indicates that something went majorly wrong during locking
+        *pid = 0;
+
+        log!(
+            Level::Debug,
+            "FsLock: trying to lock `{}'",
+            new_lock_path.display()
+        );
+
+        // first we open (and possibly create) the lock file
+        let mut openoptions = OpenOptions::new();
+        openoptions.read(true).write(true).create(true);
```

Comment:
The original code also set a mode on the file. I think this can be done with `.mode(0o600)` (from [`std::os::unix::fs::OpenOptionsExt` trait](https://doc.rust-lang.org/std/os/unix/fs/trait.OpenOptionsExt.html#tymethod.mode)).

Reply:
Nice, I didn't know about that. Thanks.